awx/main/tasks/facts.py

```diff
@@ -42,15 +30,14 @@ def start_fact_cache(inventory, destination, log_data, timeout=None, slice_numbe
 
     if timeout is None:
         timeout = settings.ANSIBLE_FACT_CACHE_TIMEOUT
-    if timeout > 0:
-        # exclude hosts with fact data older than `settings.ANSIBLE_FACT_CACHE_TIMEOUT seconds`
-        timeout = now() - datetime.timedelta(seconds=timeout)
-        hosts = _get_inventory_hosts(inventory, slice_number, slice_count, ansible_facts_modified__gte=timeout)
-    else:
-        hosts = _get_inventory_hosts(inventory, slice_number, slice_count)
+
+    if isinstance(hosts, QuerySet):
+        hosts = hosts.iterator()
 
     last_filepath_written = None
     for host in hosts:
+        if host.ansible_facts_modified and (host.ansible_facts_modified >= now() - datetime.timedelta(seconds=timeout)):
```

Comment:
Just to logic through this...

So if I understand right, facts should be expired if now() is greater than the modified timestamp + the timeout. This would mean the current time exceeds the stored timestamp plus the delta.

now() >= modified + timeout
now() - timeout >= modified

So I think the code has a sign error?

I think in the old code, the timeout was being used to determine *inclusion* of hosts (include every host that has *unexpired* facts), but here we're determining which hosts to *skip*/continue past, so we want to be true when they are expired.

But also it's 2am, and I might just not be mathing properly.